### PR TITLE
hydra: refactor pmi message handling

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description
Goals of this PR:
 * Making the code logic that deals with wire protocols and semantic protocols orthogonal
 * Making the protocol logic (wire or semantic) orthogonal to functionality
 * Use the better macros to make code more succinct
 
UPDATE:
* split cleanup commits into https://github.com/pmodels/mpich/pull/6104
* import `pmi` (just as `mpl`)
* use the wire routines from `pmi/src/pmi_wire.h`.

[skip warnings]

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [ ] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/warnings/auto

**Comment by hzhou** [comment]
test:mpich/warnings/auto

**Comment by hzhou** [comment]
test:mpich/ch3/tcp

**Comment by hzhou** [comment]
test:mpich/ch3/tcp

**Comment by hzhou** [comment]
test:mpich/ch3/tcp

**Comment by hzhou** [comment]
test:mpich/ch4/pmi

**Comment by hzhou** [comment]
test:mpich/ch4/pmi

**Comment by hzhou** [comment]
test:mpich/ch4/pmi

**Comment by hzhou** [comment]
test:mpich/ch4/pmi

**Comment by hzhou** [comment]
test:mpich/pmi

**Comment by hzhou** [comment]
test:mpich/pmi

**Comment by hzhou** [comment]
test:mpich/pmi

**Comment by hzhou** [comment]
test:mpich/ch4/pmi

**Comment by hzhou** [comment]
test:mpich/ch4/pmi

**Comment by hzhou** [comment]
test:mpich/ch4/pmi

**Comment by hzhou** [comment]
test:mpich/ch4/pmi

**Comment by hzhou** [comment]
test:mpich/ch4/pmi

**Comment by hzhou** [comment]
test:mpich/ch4/pmi

**Comment by hzhou** [comment]
test:mpich/pmi

**Review by raffenet — CHANGES_REQUESTED** [review]
6000 lines of diff over 25 commits. Please split this into smaller pieces.

File: src/util/mpir_pmi.c [hunk 1]
```diff
@@ -549,7 +549,7 @@ static int put_ex(const char *key, const void *buf, int bufsize, int is_local)
         MPIR_ERR_CHECK(mpi_errno);
         for (int i = 0; i < num_segs; i++) {
             char seg_key[50];
-            sprintf(seg_key, "%s-seg-%d/%d", key, i + 1, num_segs);
```
Comment (raffenet) [inline src/util/mpir_pmi.c:552]: NAK: this is a backwards incompatible change

**Comment by hzhou** [comment]
> 6000 lines of diff over 25 commits. Please split this into smaller pieces.

Sure.

File: src/util/mpir_pmi.c (hunk 1 above)
Comment (hzhou) [reply src/util/mpir_pmi.c:552]: We can just allow `/` in the key. I'll make the change

**Comment by hzhou** [comment]
test:mpich/pmi

**Comment by hzhou** [comment]
test:mpich/pmi

**Comment by hzhou** [comment]
test:mpich/pmi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi
test:mpich/pmi

File: src/pm/hydra/mpiexec/pmiserv_kvs.c [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+
+    status = HYD_pmiserv_pmi_reply(fd, pid, &pmi_response);
+    HYDU_ERR_POP(status, "send command failed\n");
+
+    if (cur_epoch == pg_scratch->epoch) {
+        pg_scratch->fence_count++;
+        if (pg_scratch->fence_count % proxy->pg->pg_process_count == 0) {
+            /* Poke the progress engine before exiting */
+            status = check_pending_reqs(NULL);
+            HYDU_ERR_POP(status, "poke progress error\n");
```
Comment (raffenet) [inline src/pm/hydra/mpiexec/pmiserv_kvs.c:251]:
```suggestion
            HYDU_ERR_POP(status, "check pending requests error\n");
```

**Review by raffenet — COMMENTED** [review]
Looks good. One minor suggestion. Is this the last in this series or is there more?

**Comment by hzhou** [comment]
> Is this the last in this series or is there more?

Yeah, this is the last of the previous big PR, but it is the base of new rounds of work :) Plan --
* factor out the message protocol. The current code dealing with key names, setting and checking "rc", "thrid" etc can be unified and handled by the utility layer
* Work toward reuse proxy for spawning and tree-launch - https://github.com/pmodels/mpich/pull/5943
* Add binary keyval extension
* Add minimum PMIx in hydra and libpmi, just the set that MPICH currently uses.

**Comment by raffenet** [comment]
> > Is this the last in this series or is there more?
> 
> Yeah, this is the last of the previous big PR, but it is the base of new rounds of work :) Plan --

OK, I see.

> 
> * factor out the message protocol. The current code dealing with key names, setting and checking "rc", "thrid" etc can be unified and handled by the utility layer
> * [hydra: WIP toward reuse proxy for spawning and tree-launch #5943](https://github.com/pmodels/mpich/pull/5943)
> * Add binary keyval extension
> * Add minimum PMIx in hydra and libpmi, just the set that MPICH currently uses.

I thought we had agreed that I would implement the PMIx support. I hope that is still the case.

**Comment by raffenet** [comment]
One concern I have while browsing the new code is that proxy/server functions that support only PMI1 or PMI2 are now difficult to distinguish from functions that support both. I'm not sure this is actually desirable.

File: src/pm/hydra/mpiexec/pmiserv_kvs.c (hunk 2 above)
Comment (hzhou) [reply src/pm/hydra/mpiexec/pmiserv_kvs.c]: fixed.

**Review by raffenet — APPROVED** [review]

**Pull request merged**